packages/api/src/objectSet/ObjectSet.ts

```diff
@@ -142,9 +144,48 @@ export interface InterfaceObjectSet<
 > extends MinimalObjectSet<Q> {
 }
 
+export type ExtractRdp<
+  D extends ObjectSet<any, any> | Record<string, SimplePropertyDef>,
+> = D extends Record<string, SimplePropertyDef> ? D : {};
+
+export namespace ObjectSet {
+  export namespace Fns {
+    export interface WithProperties<
+      Q extends ObjectOrInterfaceDefinition = any,
+      D extends Record<string, SimplePropertyDef> = {},
+    > {
```

Comment:
This needs to be split out so we can recombine multiple calls to `withProperties` without insane hoop jumping